### ISSUE TITLE
Recompile time with torch.compile 2.10

### ISSUE THREAD

**Issue by GLivshits** [pytorch/pytorch#174704]
Hi. After upgrading to torch 2.10 I witness heavy issues with recompilations time.
Those recompilations occur due to:
1) FSDP wrapping of model.
2) Dynamic shapes (I have 8 shapes in total).

Before upgrade warmup for all 8 shapes took 30 minutes, but after upgrade it rockets to several hours + inadequate GPU RAM overhead.

Compile regime: just usual model.compile(mode=default)

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @chauhang @mori360 @pragupta @ppwwyyxx @penguinwu @ezyang @bobrenjc93 @aditvenk @laithsakka @oulgen @jamesjwu @aorenste @anijain2305 @masnesral @coconutruben

**Comment by GLivshits** [comment]
Added this code to main script
```
import torch
torch._logging.set_logs(dynamo=logging.INFO, inductor=logging.INFO)
```

After 1.5 hours I observe a similar logs.
Logs sample:

> [rank0]:I0210 20:26:04.042000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval Eq(s100, s132) [guard added] return F.linear(input, self.weight, self.bias)  # nn/modules/linear.py:134 in forward (_meta_registrations.py:2272 in meta_mm), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Eq(s100, s132)"
> [rank0]:I0210 20:26:04.044000 12259 torch/fx/experimental/symbolic_shapes.py:6904] [8/47] set_replacement s132 = s100 (solve) VR[2, int_oo]
> [rank0]:I0210 20:26:04.059000 12259 torch/fx/experimental/symbolic_shapes.py:5225] [8/47] create_symbol s134 = 256 for L['self']._modules['1']._modules['transformer_blocks']._modules['3']._modules['attn']._modules['qk_norm'].weight.size()[0] [2, int_oo] return FusedRMSNormAffineFunction.apply(  # /home/user/model/rmsnorm.py:310 in forward (_dynamo/variables/builder.py:3702 in <lambda>), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="s134" or to suppress this message run with TORCHDYNAMO_EXTENDED_ADVICE="0"
> [rank0]:I0210 20:26:04.061000 12259 torch/fx/experimental/symbolic_shapes.py:5225] [8/47] create_symbol s135 = 6041088 for L['self']._modules['1']._modules['transformer_blocks']._modules['3']._modules['attn']._modules['qk_norm'].weight.storage_offset() [2, int_oo] return FusedRMSNormAffineFunction.apply(  # /home/user/model/rmsnorm.py:310 in forward (_dynamo/variables/builder.py:3702 in <lambda>), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="s135" or to suppress this message run with TORCHDYNAMO_EXTENDED_ADVICE="0"
> [rank0]:I0210 20:26:04.066000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval [guard suppressed] 2*s134 + 2*s135 <= 2*s73 [guard added] return FusedRMSNormAffineFunction.apply(  # /home/user/model/rmsnorm.py:310 in forward (utils/_stats.py:29 in wrapper), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="2*s134 + 2*s135 <= 2*s73"
> [rank0]:I0210 20:26:04.068000 12259 torch/fx/experimental/symbolic_shapes.py:5225] [8/47] create_symbol s136 = 256 for L['self']._modules['1']._modules['transformer_blocks']._modules['3']._modules['attn']._modules['qk_norm'].normalized_shape [-int_oo, int_oo] return FusedRMSNormAffineFunction.apply(  # /home/user/model/rmsnorm.py:310 in forward (_dynamo/variables/builder.py:2501 in wrap_symint), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="s136" or to suppress this message run with TORCHDYNAMO_EXTENDED_ADVICE="0"
> [rank0]:I0210 20:26:04.076000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval Eq(s136, 256) [guard added] output, invvar = torch.ops.rms_norm._rms_norm_affine_forward(  # /home/user/model/rmsnorm.py:251 in forward (_ops.py:1209 in __call__), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Eq(s136, 256)"
> [rank0]:I0210 20:26:04.077000 12259 torch/fx/experimental/symbolic_shapes.py:6904] [8/47] set_replacement s136 = 256 (range_refined_to_singleton) VR[256, 256]
> [rank0]:I0210 20:26:04.117000 12259 torch/fx/experimental/symbolic_shapes.py:5225] [8/47] create_symbol s137 = 4 for L['self']._modules['1']._modules['transformer_blocks']._modules['3']._modules['attn'].n_heads [-int_oo, int_oo] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (_dynamo/variables/builder.py:2501 in wrap_symint), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="s137" or to suppress this message run with TORCHDYNAMO_EXTENDED_ADVICE="0"
> [rank0]:I0210 20:26:04.120000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval Ne(s137, -1) [guard added] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (_prims_common/__init__.py:1051 in infer_size), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Ne(s137, -1)"
> [rank0]:I0210 20:26:04.121000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval s137 >= 0 [guard added] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (_prims_common/__init__.py:1055 in infer_size), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="s137 >= 0"
> [rank0]:I0210 20:26:04.126000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval Eq(4032*s125, 4032*s137*((s125//s137))) [guard added] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (_prims_common/__init__.py:1064 in infer_size), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Eq(4032*s125, 4032*s137*((s125//s137)))"
> [rank0]:I0210 20:26:04.133000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval Ne(s137, s125) [guard added] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (_refs/__init__.py:3897 in _reshape_view_helper_core_alg), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Ne(s137, s125)"
> [rank0]:I0210 20:26:04.134000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval Eq(Mod(s125, s137), 0) [guard added] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (_refs/__init__.py:3903 in _reshape_view_helper_core_alg), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Eq(Mod(s125, s137), 0)"
> [rank0]:I0210 20:26:04.996000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval Ne((s125//s137), 0) [guard added] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (_prims_common/__init__.py:2016 in are_strides_like_channels_last_or_false), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Ne((s125//s137), 0)"
> [rank0]:I0210 20:26:04.999000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval [guard suppressed] Ne(s137, 0) [guard added] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (utils/_stats.py:29 in wrapper), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Ne(s137, 0)"
> [rank0]:I0210 20:26:05.000000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval [guard suppressed] Ne((s125//s137), 0) [guard added] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (utils/_stats.py:29 in wrapper), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Ne((s125//s137), 0)"
> [rank0]:I0210 20:26:05.002000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval [guard suppressed] Ne(4032*s137*((s125//s137)), 0) [guard added] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (utils/_stats.py:29 in wrapper), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Ne(4032*s137*((s125//s137)), 0)"
> [rank0]:I0210 20:26:05.003000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval [guard suppressed] Ne(4032*s137*((s125//s137)), 0) [guard added] q = rearrange(q, "b n (h d) -> b h n d", h=self.n_heads)  # /home/user/model/attention.py:694 in scaled_dot_prod_attn_op (utils/_stats.py:29 in wrapper), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Ne(4032*s137*((s125//s137)), 0)"

**Comment by laithsakka** [comment]
specializaion happens
[rank0]:I0210 20:26:04.076000 12259 torch/fx/experimental/symbolic_shapes.py:7292] [8/47] eval Eq(s136, 256) [guard added] output, invvar = torch.ops.rms_norm._rms_norm_affine_forward( # /home/user/model/rmsnorm.py:251 in forward (_ops.py:1209 in call), for more info run with TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Eq(s136, 256)"
run with 
 TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Eq(s136, 256)" to see the reason of specialization. 

Note: 
mark input  size unbacked if you want to guarantee no specialization happen to it.

**Comment by laithsakka** [comment]
> > This seems fixed, do we need this open? [@GLivshits](https://github.com/GLivshits)
> > In general if you want a guarantee for no recompilations you should use unbacked dynamic shapes. you might DDEs if you do so but then you know would have recompiled if you dont hit those, form there you can use guard_or_ or other documented techniques to bypass.
> 
> 1. Will this fix be included in 2.11 release?
> 2. I've checked plenty of resources, including two google docs with torch.compile tutorial, and did not see references with guard_or. Can you please share those resources?
>    I use transformer with packed sequences (batch of [seqlen_sum, dim]), and dynamism is unavoidable in such case.

did you try to mark your inputs unbacked using mark_unbacked?
